### ISSUE TITLE
Trial Validator

### ISSUE THREAD

**Issue by TysonKerr** [gikeymarcia/Collector#126]
I had an idea for validating trials at login. Each trial can have another file called validator.php, which has some code like
```php
    return function($inputs) {
        // do some error checking here
    };
```

for example, if you had a trial that needed a "Cue" column, a "Number" column with a numeric value, and a color setting inside the "Settings" column, you could write
```php
    return function($inputs) {
        $errors = array();
        if (!isset($inputs['Cue'])) {
            $errors[] = '"Cue" column missing, required for ' . $trialType;
        }
        
        if (!isset($inputs['Number'])) {
            $errors[] = '"Number" column missing, required for ' . $trialType;
        } elseif (!is_numeric($inputs['Number'])) {
            $errors[] = '"Number" must be a numeric value, ' 
                      . $inputs['Number'] . ' given';
        }
        
        $settings = explode($input['Settings']);
        $hasColor = false;
        
        foreach ($settings as $setting) {
            $keyValuePair = explode('=', $setting);
            $key = trim($keyValuePair[0]);
            $key = strtolower($key);
            
            if ($key === 'color') {
                $hasColor = true;
                break;
            }
        }
        
        if (!$hasColor) {
            $errors[] = 'Setting missing: "color" <br>'
                      . 'This must be present in the "Settings" column';
        }
        
        return $errors;
    };
```

Then at login, after getting the stim and proc files stitched together, but before shuffling, you could run through each row like so:
```php
    $errors     = array();
    $validators = array();
    $trialTypes = getAllTrialTypes();
    foreach ($procedure as $pos => $row) {
        foreach ($postTrial as $postN) {
            // get all the stim and proc columns for this post trial
            $inputs = getTrialValues($row, $postN, $stimuli);
            
            $trialType = $inputs['Trial Type'];
            
            // find validating function, if exists
            if (!isset($validators[$trialType])) {
                if (isset($trialTypes[$trialType]['Validator'])) {
                    $validators[$trialType] = require $trialTypes[$trialType]['Validator'];
                } else {
                    $validators[$trialType] = false;
                }
            }
            
            // run validating function, if exists
            if ($validators[$trialType] !== false) {
                $errors["$pos.$postN"] = $validators[$trialType]($inputs);
            }
        }
    }
    
    $hasError = false;
    foreach ($errors as $trial => $theseErrors) {
        if ($theseErrors !== array()) {
            $hasError = true;
            echo "<div>Error for trial $trial</div>";
            foreach ($theseErrors as $thisError) {
                echo $thisError;
            }
        }
    }
    
    if ($hasError) {
        exit;
    }
```

**Comment by TysonKerr** [comment]
here is the function `getTrialValues()`, for extracting any postTrial out of the procedure and stimuli file:
```php
function getTrialValues($procRow, $postN, $stimuli) {
    $procCols = array();
    $procCols['Item'] = $procRow['Item'];
    if ($postN === 0) {
        foreach ($procRow as $col => $val) {
            if (substr($col, 0, 5) === 'Post ') {
                continue;
            } else {
                $procCols[$col] = $val;
            }
        }
    } else {
        $colPre = 'Post ' . $postN . ' ';
        $colPreLen = strlen($colPre);
        foreach ($procRow as $col => $val) {
            if (substr($col, 0, $colPreLen) !== $colPre) {
                continue;
            } else {
                $colClean = substr($col, $colPreLen);
                $procCols[$colClean] = $val;
            }
        }
    }
    
    $items = rangeToArray($procCols['Item']);
    $theseStim = array();
    foreach ($items as $i) {
        if (!isset($stimuli[$i]) || $stimuli[$i] === 0) {
            continue;
        } else {
            $theseStim[] = $stimuli[$i];
        }
    }
    
    if ($theseStim === array()) {
        $stimHeaders = array_keys($stimuli[0]);
        foreach ($stimHeaders as $stimCol) {
            $theseStim[0][$stimCol] = 'na';
        }
    }
    
    $stimMerged = array();
    foreach ($theseStim as $row) {
        foreach ($row as $col => $val) {
            $stimMerged[$col][] = $val;
        }
    }
    
    foreach ($stimMerged as $col => $vals) {
        $stimMerged[$col] = implode('|', $vals);
    }
    
    $inputs = $stimMerged;
    // merging manually, because array_merge will re-number numerical keys
    foreach ($procCols as $col => $val) {
        $inputs[$col] = $val;
    }
    
    return $inputs;
}
```
I feel like I've written this function several times now. . . . But this should basically be all we need to run this validation. Then we can add efficiency checks, like checking the edit times of these files, and if we have already scanned everything and no files have been edited, then we can skip this step. But this would be a good step towards purposeful error messages, rather than just letting Collector crash and burn when something goes wrong.

On the topic of checking file edit times, it might actually be pretty simple to write a function that scans the entire code and Experiment folder, gets the file edit times, joins those into a string, and then hashes it, and then if we log the hash every time we do these error checks, then we can confidently skip the error checks for the next 100 participants, because nothing will have changed except the data folder. That way, we can right super elaborate error checks, which takes several seconds to run, but as soon as we get through all the checks without errors, then the fileEdit check just spends a few milliseconds at login making sure that nothing has changed.

**Comment by TysonKerr** [comment]
Note: need to check if post trial is "off"

**Comment by adamblake** [comment]
This feature was implemented in #162
